File: app/src/main/java/eu/kanade/tachiyomi/ui/setting/SettingsAdvancedFragment.java
package eu.kanade.tachiyomi.ui.setting;

import android.os.Bundle;
import android.preference.Preference;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.afollestad.materialdialogs.MaterialDialog;

import java.io.File;
import java.util.concurrent.atomic.AtomicInteger;

import eu.kanade.tachiyomi.R;
import eu.kanade.tachiyomi.data.cache.ChapterCache;
import eu.kanade.tachiyomi.data.database.DatabaseHelper;
import eu.kanade.tachiyomi.util.ToastUtil;
import rx.Observable;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;
import rx.subscriptions.CompositeSubscription;

public class SettingsAdvancedFragment extends SettingsNestedFragment {

    private CompositeSubscription subscriptions;

    public static SettingsNestedFragment newInstance(int resourcePreference, int resourceTitle) {
        SettingsNestedFragment fragment = new SettingsAdvancedFragment();
        fragment.setArgs(resourcePreference, resourceTitle);
        return fragment;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedState) {
        View view = super.onCreateView(inflater, container, savedState);
        subscriptions = new CompositeSubscription();

        Preference clearCache = findPreference(getString(R.string.pref_clear_chapter_cache_key));
        clearCache.setOnPreferenceClickListener(preference -> {
            clearChapterCache(preference);
            return true;
        });
        clearCache.setSummary(getString(R.string.used_cache, getChapterCache().getReadableSize()));

        Preference clearDatabase = findPreference(getString(R.string.pref_clear_database_key));
        clearDatabase.setOnPreferenceClickListener(preference -> {
            clearDatabase();
            return true;
        });

        return view;
    }

    @Override
    public void onDestroyView() {
        subscriptions.unsubscribe();
        super.onDestroyView();
    }

    private void clearChapterCache(Preference preference) {
        final ChapterCache chapterCache = getChapterCache();
        final AtomicInteger deletedFiles = new AtomicInteger();

        File[] files = chapterCache.getCacheDir().listFiles();

        MaterialDialog dialog = new MaterialDialog.Builder(getActivity())
                .title(R.string.deleting)
                .progress(false, files.length, true)
                .cancelable(false)
                .show();

        subscriptions.add(Observable.defer(() -> Observable.from(files))
                .concatMap(file -> {
                    if (chapterCache.removeFileFromCache(file.getName())) {
                        deletedFiles.incrementAndGet();
                    }
                    return Observable.just(file);
                })
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(file -> dialog.incrementProgress(1),
                        error -> {
                            dialog.dismiss();
                            ToastUtil.showShort(getActivity(), getString(R.string.cache_delete_error));
                        }, () -> {
                            dialog.dismiss();
                            ToastUtil.showShort(getActivity(), getString(R.string.cache_deleted, deletedFiles.get()));
                            preference.setSummary(getString(R.string.used_cache, chapterCache.getReadableSize()));
                        }));
    }

    private void clearDatabase() {
        final DatabaseHelper db = getSettingsActivity().db;

        new MaterialDialog.Builder(getActivity())
                .content(R.string.clear_database_confirmation)
                .positiveText(R.string.button_yes)
                .negativeText(R.string.button_no)
                .onPositive((dialog1, which) -> {
                    db.deleteMangasNotInLibrary().executeAsBlocking();
                })
                .show();
    }

    private ChapterCache getChapterCache() {
        return getSettingsActivity().chapterCache;
    }

}

File: app/src/main/java/eu/kanade/tachiyomi/ui/setting/SettingsAdvancedFragment.kt
package eu.kanade.tachiyomi.ui.setting

import android.os.Bundle
import android.support.v7.preference.Preference
import android.support.v7.preference.XpPreferenceFragment
import android.view.View
import com.afollestad.materialdialogs.MaterialDialog
import eu.kanade.tachiyomi.R
import eu.kanade.tachiyomi.data.cache.ChapterCache
import eu.kanade.tachiyomi.data.database.DatabaseHelper
import eu.kanade.tachiyomi.data.library.LibraryUpdateService
import eu.kanade.tachiyomi.data.network.NetworkHelper
import eu.kanade.tachiyomi.util.plusAssign
import eu.kanade.tachiyomi.util.toast
import rx.Observable
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import uy.kohesive.injekt.injectLazy
import java.util.concurrent.atomic.AtomicInteger

class SettingsAdvancedFragment : SettingsFragment() {

    companion object {
        fun newInstance(rootKey: String): SettingsAdvancedFragment {
            val args = Bundle()
            args.putString(XpPreferenceFragment.ARG_PREFERENCE_ROOT, rootKey)
            return SettingsAdvancedFragment().apply { arguments = args }
        }
    }

    private val network: NetworkHelper by injectLazy()

    private val chapterCache: ChapterCache by injectLazy()

    private val db: DatabaseHelper by injectLazy()

    private val clearCache: Preference by bindPref(R.string.pref_clear_chapter_cache_key)

    private val clearDatabase: Preference by bindPref(R.string.pref_clear_database_key)

    private val clearCookies: Preference by bindPref(R.string.pref_clear_cookies_key)

    private val refreshMetadata: Preference by bindPref(R.string.pref_refresh_library_metadata_key)

    override fun onViewCreated(view: View, savedState: Bundle?) {
        super.onViewCreated(view, savedState)

        clearCache.setOnPreferenceClickListener {
            clearChapterCache()
            true
        }
        clearCache.summary = getString(R.string.used_cache, chapterCache.readableSize)

        clearCookies.setOnPreferenceClickListener {
            network.cookies.removeAll()
            activity.toast(R.string.cookies_cleared)
            true
        }

        clearDatabase.setOnPreferenceClickListener {
            clearDatabase()
            true
        }

        refreshMetadata.setOnPreferenceClickListener {
            LibraryUpdateService.start(context, details = true)
            true
        }
    }

    private fun clearChapterCache() {
        val deletedFiles = AtomicInteger()

        val files = chapterCache.cacheDir.listFiles() ?: return

        val dialog = MaterialDialog.Builder(activity)
                .title(R.string.deleting)
                .progress(false, files.size, true)
                .cancelable(false)
                .show()

        subscriptions += Observable.defer { Observable.from(files) }
                .concatMap { file ->
                    if (chapterCache.removeFileFromCache(file.name)) {
                        deletedFiles.incrementAndGet()
                    }
                    Observable.just(file)
                }
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({
                    dialog.incrementProgress(1)
                }, {
                    dialog.dismiss()
                    activity.toast(R.string.cache_delete_error)
                }, {
                    dialog.dismiss()
                    activity.toast(getString(R.string.cache_deleted, deletedFiles.get()))
                    clearCache.summary = getString(R.string.used_cache, chapterCache.readableSize)
                })
    }

    private fun clearDatabase() {
        MaterialDialog.Builder(activity)
                .content(R.string.clear_database_confirmation)
                .positiveText(android.R.string.yes)
                .negativeText(android.R.string.no)
                .onPositive { dialog, which ->
                    (activity as SettingsActivity).parentFlags = SettingsActivity.FLAG_DATABASE_CLEARED
                    db.deleteMangasNotInLibrary().executeAsBlocking()
                    activity.toast(R.string.clear_database_completed)
                }
                .show()
    }

}
